build-tools/src/main/java/org/elasticsearch/gradle/testclusters/MockApmServer.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.gradle.testclusters;
+
+import com.sun.net.httpserver.HttpExchange;
+import com.sun.net.httpserver.HttpHandler;
+import com.sun.net.httpserver.HttpServer;
+
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.InetSocketAddress;
+
+/**
+ * This is a server which just accepts lines of JSON code and if the JSON
+ * is valid and the root node is "transaction", then adds that JSON object
+ * to a transaction list which is accessible externally to the class.
+ *
+ * The Elastic agent sends lines of JSON code, and so this mock server
+ * can be used as a basic APM server for testing.
+ *
+ * The HTTP server used is the JDK embedded com.sun.net.httpserver
+ */
+public class MockApmServer {
+    private static final Logger logger = Logging.getLogger(MockApmServer.class);
+    private int port;
+
+    public MockApmServer(int port) {
+        this.port = port;
+    }
+
+    /**
+     * Simple main that starts a mock APM server and prints the port it is
+     * running on. This is not needed
+     * for testing, it is just a convenient template for trying things out
+     * if you want play around.
+     */
+    public static void main(String[] args) throws IOException, InterruptedException {
+        MockApmServer server = new MockApmServer(9999);
+        server.start();
+    }
+
+    private static volatile HttpServer instance;
+
+    /**
+     * Start the Mock APM server. Just returns empty JSON structures for every incoming message
+     * @return - the port the Mock APM server started on
+     * @throws IOException
+     */
+    public synchronized int start() throws IOException {
+        if (instance != null) {
+            throw new IOException("MockApmServer: Ooops, you can't start this instance more than once");
+        }
+        InetSocketAddress addr = new InetSocketAddress("0.0.0.0", port);
+        HttpServer server = HttpServer.create(addr, 10);
+        server.createContext("/exit", new ExitHandler());
```

Comment:
Rather than require an explicit exit call, can we make this process self-closing by breaking when the parent process exits? That is, do a read from stdin, that will block indefinitely, and when the pipe breaks (the parent exited), this process exits.